### DISCUSSION TITLE
Smart Crop to Detail

### DISCUSSION THREAD

**Discussion by wave-dash** [ImageMagick/ImageMagick#6171] [Help]
### ImageMagick version

7.1.1-Q16-HDRI

### Operating system, version and so on

Windows 11 Pro

### Description

I'm trying to use ImageMagick to intelligently (or reasonably intelligently) crop an image to the most complex/detailed part of the image.

Someone had a similar question way back in 2017: https://www.imagemagick.org/discourse-server/viewtopic.php?t=32232

User snibgo seems to have written a fairly complex solution for this: http://im.snibgo.com/crop2det.htm

However, I can't get it to work. As a result of me being unfamiliar with ImageMagick and these scripts relying on a bunch of custom dependencies, I can't tell what's going wrong.

```
C:\Users\Name\Documents\imsnibgoBats>cropToDetailSml.bat bsd.png out.png 500x500

C:\Users\Name\Documents\imsnibgoBats>rem From bsd.png an ordinary image,

C:\Users\Name\Documents\imsnibgoBats>rem makes out.png cropped to detail of bsd.png.

C:\Users\Name\Documents\imsnibgoBats>rem 500x500 is required dimensions, WxH.

C:\Users\Name\Documents\imsnibgoBats>rem  force the use of this threshold value.
cropToDetailSml.bat: bsd.png out.png 500x500
cropToDetailSml.bat: ctdBLR_SIG= ctdUSE_CONN_COMP= ctdDO_RECURSE=
cropToDetailSml.bat: WW=760 HH=773 req:500x500 ctdBLR_SIG=0 ctdUSE_CONN_COMP=1 ctdDO_RECURSE=1 DIAG_FACT=1.63258 FORCE_TH=
cropToDetailSml.bat: gsm BLR=0.81629
convert: unable to open image '\temp\gsm_knl.miff': No such file or directory @ error/blob.c/OpenBlob/3569.
convert: Empty expression in parentheses at '*2+1==h?1:0' @ error/fx.c/GetOperand/2089.
convert: Empty expression in parentheses at '*2+1==h?1:0' @ error/fx.c/GetOperand/2089.
```

```
C:\Users\Name\Documents\imsnibgoBats>cropToDetailLge.bat bsd.png out.png 500x500

C:\Users\Name\Documents\imsnibgoBats>rem Like cropToDetailSml.bat, but for large images.
cropToDetailLge.bat: bsd.png out.png 500x500
cropToDetailLge.bat: ctdBLR_SIG= ctdUSE_CONN_COMP= ctdDO_RECURSE=
cropToDetailLge.bat: WW=760 HH=773 req:500x500 ctdBLR_SIG= ctdUSE_CONN_COMP= ctdDO_RECURSE= DIAG_FACT= FORCE_TH=
magick: unable to open image '\temp\ctdl_sml.miff': No such file or directory @ error/blob.c/OpenBlob/3569.
```

I was able to get this primitive version of the script to work, though: https://www.imagemagick.org/discourse-server/viewtopic.php?p=147478#p147478. With the same code and same dragonfly image, I was able to get the same cropped output, which is pretty promising. Can anyone help me with the full scripts?

**Comment by snibgo** [comment]
> unable to open image '\temp\gsm_knl.miff'

Have you checked that you have a directory named `\temp`? If you don't, then create it and try again.

Note that gaussSlpMag.bat uses chStrs, which is not published. Instead, use "sed" to change strings in the files.

**Reply by wave-dash** [reply]
Adding a temp folder (in the same directory as the .bat files, I assume) doesn't change anything.

Could you elaborate on using sed? I assume it isn't as simple as swapping out the commands?

**Comment by snibgo** [comment]
The leading backslash means it should be a subdirectory of the root. This command will make the directory:
```
md \temp
```
> I assume it isn't as simple as swapping out the commands?

It is that simple.

In `gaussSlpMag.bat`, instead of the four calls to chStrs, you could use:
```
sed -e s/:/+%OFFS%+%OFFS%:/ --in-place=BAK %KNL_L%
sed -e s/:/+0+%OFFS%:/ --in-place=BAK %KNL_R%
sed -e s/:/+%OFFS%+%OFFS%:/ --in-place=BAK %KNL_T%
sed -e s/:/+%OFFS%+0:/ --in-place=BAK %KNL_B%
```

**Comment by fmw42** [comment]
If you can use Windows Unix or Cygwin, you can try my Unix bash scripts:  smartcrop, smarttrim, and trim2detail at http://www.fmwconcepts.com/imagemagick/index.php
